Convert double[][] -> Img & other fun stuff

gselzer opened the pull request:
This PR is designed to provide developers with an informative walkthrough in expanding SciJava Ops' conversion capabilities via adding new converters.

There's a bunch of leftover cruft that should be removed if others like the general direction of this PR.  Notably, I converted the `Purpose` document to ReStructuredText so that I could add a reference label for easy cross-document references.

hinerm commented on docs/ops/doc/ParameterConversion.rst:
```diff
@@ -0,0 +1,141 @@
+This Op might work well, however if users have a small kernel that is *only* used for this Op, they may find it frustrating to represent that data as a ``RandomAccessibleInterval``. Fortunately, SciJava Ops allows us to write ``engine.convert`` Ops, so users can pass their data in a data structure more convenient for them.
+
+
+.. code-block:: java
```
This code block would benefit from more explanation. This is what would be required WITHOUT a converter, correct?

The switch from `double[]` here to `double[][]` in the converter example could be confusing. Make it consistent.

hinerm commented on docs/ops/doc/ParameterConversion.rst:
```diff
@@ -0,0 +1,141 @@
+
+In this case, users might find it nicer to specify their kernel as a ``double[][]``, which is much easier for users to construct
+
+.. code-block:: java
```
I think I would put this block after the converter definition, to show what is now possible with the converter

hinerm commented on docs/ops/doc/ParameterConversion.rst:
```diff
@@ -0,0 +1,141 @@
+
+The only step for us as the developer is to tell SciJava Ops that it can convert ``double[][]``\ s to ``RandomAccessibleInterval<DoubleType>``\ s, which we do with ``engine.convert`` Ops.
+
+A simple ``engine.convert`` Op
```
This is nice but I think it's incomplete or potentially misleading, because an `engine.convert` Op alone is insufficient right? Don't we also need `copy` Ops?

I imagine this example works because we already have `copy` Ops for `double[][]` but presumably this would be used when adding support for new types that do not have existing `copy` Ops.

gselzer replied:
Made the section, let me know what you think!

hinerm commented on docs/ops/doc/ParameterConversion.rst:
```diff
@@ -0,0 +1,141 @@
+    Img<DoubleType> in = ...
+    // 3x3 averaging kernel
+    double m = 1/9;
+    double[] kernel = new double[][] { //
```
When making a `double[][]` you don't need the internal `new double[]`.. just `{{m, m, m}, {m, m, m}, {m, m, m}}` should be sufficient

ctrueden replied:
Also, `double[] kernel = new double[][]...` has conflicting types.

gselzer replied:
Fixed!

hinerm commented on docs/ops/doc/Concepts.md:
```diff
@@ -3,7 +3,7 @@
 This page is designed to help *Ops users* understand the SciJava Ops concepts powering the framework.
 
 ## Ops
-An **algorithm** is a mathematical routine that transforms, interrogate, or refines input values into output values. Algorithms are used throughout scientific computing, and the fundamental [purpose](Purpose) of SciJava Ops is to facilitate their application.
```
I don't know Sphinx super well but when I was playing with it locally it seemed like file extensions were not necessary for links to work. I favored leaving them absent, e.g. if we change a `.rst` to an `.md` will it break? Or will it not since it's all being built as HTML anyway, in which case why include them?

Just wondering if there's a definitive reason to include them?

gselzer replied:
Good catch - removed it!

hinerm commented on docs/ops/doc/examples/opencv_denoise.rst:
```diff
@@ -2,7 +2,7 @@
 ========================================
 
-In this example we will use a denoise algorithm from the external `OpenCV libray`_.
+In this example we will use a denoise algorithm from the external `OpenCV library`_.
```
I like to eat strawbrerries in the libary

ctrueden commented on docs/ops/doc/ParameterConversion.rst:
```diff
@@ -0,0 +1,141 @@
+Parameter Conversion for Developers
+========================================
+
+In this example, we show how Ops developers can implement parameter conversion to enable seamless interoperability between Ops utilizing different data structures.between
```
`.between` trailing

ctrueden commented on docs/ops/doc/ParameterConversion.rst:
```diff
@@ -0,0 +1,141 @@
+.. code-block:: java
+
+	/**
+	 * Conolves an image with a kernel, returning the output in a new object
```
`Conolves`

ctrueden commented on docs/ops/doc/ParameterConversion.rst:
```diff
@@ -0,0 +1,141 @@
+Basics
+======
+
+A key :ref:`value <driving-values>` of SciJava Ops is flexibility, and flexibility is (in part) achieved through **parameter conversion**. At its core, parameter conversion allows *translation* of data stored in one data structure (e.g. an ImgLib2 ``RandomAccessibleInterval``) into a different data structure (e.g. an OpenCV ``Mat``) **on the fly**. This allows SciJava Ops to execute Ops backed by ...
```
You can just say "A value" rather than "key value". All values are "key"

ctrueden commented on docs/ops/doc/ParameterConversion.rst:
```diff
@@ -0,0 +1,141 @@
+        m, m, m //
+    };
+    // Not ideal
+    Img<DoubleType> kernel = ArrayImgs.doubles(data, new long[] {3, 3});
```
You can leave off the `new long[] {` portion, no? Because it's varargs?

gselzer replied:
Uh yes, that's a remnant of my experiments with `PlanarImg`, I think

ctrueden commented on docs/ops/doc/ParameterConversion.rst:
```diff
@@ -0,0 +1,141 @@
+
+    var result = ops.op("filter.convolve").input(in, kernel).apply();
+
+In this case, users might find it nicer to specify their kernel as a ``double[][]``, which is much easier for users to construct
```
I would soften `much easier` &rarr; `easier`

ctrueden commented on docs/ops/doc/ParameterConversion.rst:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    // Ideal case - no need to wrap to Img
+    var result = ops.op("filter.convolve").input(in, kernel).apply();
```
Untyped function matches considered harmful! :smiling_imp:

gselzer replied:
If you prefer the following, I'll replace it!
```java
var result = ops.op("filter.convolve").input(in, kernel).outType(new Nil<RandomAccessibleInterval<DoubleType>>() {}).apply();
```

ctrueden commented on docs/ops/doc/ParameterConversion.rst:
```diff
@@ -0,0 +1,141 @@
+A simple ``engine.convert`` Op
+==============================
+
+All ``engine.convert`` Ops are simple ``Function``\ s, that take as input the user argument to the Op, and return a *translation* of that data into the type expected by Ops. In our case, we want to convert *from* the user's ``double[][]`` into a ``RandomAccessibleInterval<DoubleType>``:
```
Drop `simple` qualifier

ctrueden commented on docs/ops/doc/ParameterConversion.rst:
```diff
@@ -0,0 +1,141 @@
+        return img;
+    }
+
+This Op, discovered through the SciJava Ops Indexer, is **all** that is needed to make the execution pattern we want functional.
```
Bolding the word "all", and then later saying well actually, you might also need an `engine.copy` Op if you want to use computers, seems disingenuous to me.

ctrueden commented on docs/ops/doc/ParameterConversion.rst:
```diff
@@ -0,0 +1,141 @@
+				input[row][col] = value;
+			}
+		};
+		return ArrayImgs.doubles(access, input.length, input[0].length);
```
This is very slick :sunglasses:

gselzer replied:
Yeah, I was really happy with it - only issue is that in practice it would fail for `double[][]`s with many elements due to too many positions.

ctrueden commented on docs/ops/doc/Purpose.rst:
```diff
@@ -1,50 +1,68 @@
-# Purpose
+=======
+Purpose
+=======
```
Super Bikeshed Man wonders: Are we changing all headers from `#` style over to `=====`/`------` now? Or does `#` style not work in .rst? Either way: should the .md files not then be consistent with this style?

gselzer replied:
Yeah, I'm pretty sure `#` means something different within rst [sections](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#sections)

gselzer commented:
@hinerm @ctrueden thank you so much for the review! I think I've addressed all your points, and have also written detailed sections on output conversion (for both `Function`s and `Computer`s/`Inplace`s). Let me know what you think!
